glommio/src/io/directory.rs

```diff
@@ -105,3 +136,12 @@ impl Directory {
         self.file.close().await
     }
 }
+
+fn contains_dir(path: &Path) -> bool {
+    let mut iter = path.components().peekable();
+    let is_file = match iter.peek() {
+        Some(std::path::Component::Normal(_)) => true,
+        _ => false,
+    };
+    iter.count() > 1 || !is_file
```

Comment:
The `iter.count() > 1` check should be moved inside the appropriate branch of the `match` statement (`Some(std::path::Component::Normal(_))`). Then, the iterator doesn't need to be `peekable`, because if it wouldn't be peekable, only the `iter.count() > 1` check would change to `iter.count() > 0`, which is equivalent to `iter.next().is_some()`.

Code suggestion:
```rust
fn contains_dir(path: &Path) -> bool {
    let mut iter = path.components();
    match iter.next() {
        Some(std::path::Component::Normal(_)) => iter.next().is_some(),
        _ => true,
    }
}
```

I'm not completely sure if that suggestion is correct, this should be tested.

Reply:
yes that's much better, thanks for the suggestion! 

I'm not sure what the correct behavior should be in the case where path is `foo/`. the `components()` method would [normalizedthe trailing slash away](https://doc.rust-lang.org/std/path/struct.Path.html#method.components) and so this function would return `true`. is this what we want?

Reply:
I would expect that the "openat" equivalent should act like:
1. `chdir()` to the directory is called.
2. `File::create` / `File::open` with the given Path is called. If this call would also normalize the trailing `/` away, then this behavoir is probably what we want. If not, then we must check for that, too.

Reply:
We should not `chdir`
If we can open a file in the current directory that is represented by the object, fine. Otherwise this needs to fail.

The main thing that needs to be handle correctly is files with `/` in their name. The example you gave, `foo/` is not a valid file name.

To the best of my knowledge `/` is invalid in path names.

I fully expect `Path` to do the right thing here across OS btw: if something is a valid filename with quotes and all, and `Path::components` break it apart, I would consider it a bug.

Reply:
I know. I wrote "should *act like*". It should handle that case the same way `std::fs::OpenOptions::open` acts when given a file name with trailing slash.